File: android/designsystem/src/main/java/com/khipu/designsystem/components/KdsButton.kt
package com.khipu.designsystem.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonColors
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.khipu.designsystem.theme.KdsComponentShapes
import com.khipu.designsystem.theme.KdsTextStyles
import com.khipu.designsystem.theme.KdsTheme
import com.khipu.designsystem.tokens.KdsBorderRadius
import com.khipu.designsystem.tokens.KdsColors
import com.khipu.designsystem.tokens.KdsDimensions

/**
 * Khipu Design System - Button
 *
 * A themed button component with variants (contained, outlined, text),
 * color schemes, sizes, and loading state support.
 */

// =============================================================================
// ENUMS
// =============================================================================

/**
 * Button visual variant
 */
enum class KdsButtonVariant {
    /** Filled button with primary color background */
    CONTAINED,
    /** Button with border and transparent background */
    OUTLINED,
    /** Text-only button with no background or border */
    TEXT
}

/**
 * Button color scheme
 */
enum class KdsButtonColor {
    PRIMARY,
    SECONDARY,
    SUCCESS,
    WARNING,
    ERROR
}

/**
 * Button size
 */
enum class KdsButtonSize {
    SMALL,
    MEDIUM,
    LARGE
}

// =============================================================================
// BUTTON COMPONENT
// =============================================================================

/**
 * Khipu Design System Button
 *
 * @param text The button label text
 * @param onClick Callback when button is clicked
 * @param modifier Modifier for the button
 * @param variant Visual style variant (CONTAINED, OUTLINED, TEXT)
 * @param color Color scheme (PRIMARY, SECONDARY, SUCCESS, WARNING, ERROR)
 * @param size Button size (SMALL, MEDIUM, LARGE)
 * @param enabled Whether the button is enabled
 * @param loading Whether to show loading spinner
 * @param fullWidth Whether button should fill available width
 * @param leadingIcon Optional composable for leading icon
 * @param trailingIcon Optional composable for trailing icon
 *
 * @sample
 * ```
 * KdsButton(
 *     text = "Pay Now",
 *     onClick = { /* handle click */ },
 *     variant = KdsButtonVariant.CONTAINED,
 *     color = KdsButtonColor.PRIMARY,
 *     loading = isProcessing
 * )
 * ```
 */
@Composable
fun KdsButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: KdsButtonVariant = KdsButtonVariant.CONTAINED,
    color: KdsButtonColor = KdsButtonColor.PRIMARY,
    size: KdsButtonSize = KdsButtonSize.MEDIUM,
    enabled: Boolean = true,
    loading: Boolean = false,
    fullWidth: Boolean = true,
    leadingIcon: @Composable (() -> Unit)? = null,
    trailingIcon: @Composable (() -> Unit)? = null,
) {
    val colors = getButtonColors(variant, color)
    val contentPadding = getButtonPadding(size)
    val minHeight = getButtonHeight(size)

    val buttonModifier = modifier
        .then(if (fullWidth) Modifier.fillMaxWidth() else Modifier)
        .height(minHeight)

    val buttonContent: @Composable () -> Unit = {
        if (loading) {
            CircularProgressIndicator(
                modifier = Modifier.size(20.dp),
                color = colors.contentColor,
                strokeWidth = 2.dp
            )
        } else {
            leadingIcon?.let {
                it()
                Spacer(Modifier.width(8.dp))
            }

            Text(
                text = text,
                style = KdsTextStyles.button,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )

            trailingIcon?.let {
                Spacer(Modifier.width(8.dp))
                it()
            }
        }
    }

    when (variant) {
        KdsButtonVariant.CONTAINED -> {
            Button(
                onClick = onClick,
                modifier = buttonModifier,
                enabled = enabled && !loading,
                colors = colors,
                contentPadding = contentPadding,
                shape = KdsComponentShapes.button,
            ) {
                buttonContent()
            }
        }

        KdsButtonVariant.OUTLINED -> {
            OutlinedButton(
                onClick = onClick,
                modifier = buttonModifier,
                enabled = enabled && !loading,
                colors = colors,
                contentPadding = contentPadding,
                shape = KdsComponentShapes.button,
                border = BorderStroke(
                    width = 1.dp,
                    color = if (enabled) getColorMain(color) else KdsColors.actionDisabled
                )
            ) {
                buttonContent()
            }
        }

        KdsButtonVariant.TEXT -> {
            TextButton(
                onClick = onClick,
                modifier = buttonModifier,
                enabled = enabled && !loading,
                colors = colors,
                contentPadding = contentPadding,
                shape = KdsComponentShapes.button,
            ) {
                buttonContent()
            }
        }
    }
}

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

@Composable
private fun getButtonColors(
    variant: KdsButtonVariant,
    color: KdsButtonColor
): ButtonColors {
    val mainColor = getColorMain(color)
    val contrastColor = getColorContrast(color)

    return when (variant) {
        KdsButtonVariant.CONTAINED -> ButtonDefaults.buttonColors(
            containerColor = mainColor,
            contentColor = contrastColor,
            disabledContainerColor = KdsColors.actionDisabledBackground,
            disabledContentColor = KdsColors.actionDisabled
        )

        KdsButtonVariant.OUTLINED -> ButtonDefaults.outlinedButtonColors(
            containerColor = Color.Transparent,
            contentColor = mainColor,
            disabledContainerColor = Color.Transparent,
            disabledContentColor = KdsColors.actionDisabled
        )

        KdsButtonVariant.TEXT -> ButtonDefaults.textButtonColors(
            containerColor = Color.Transparent,
            contentColor = mainColor,
            disabledContainerColor = Color.Transparent,
            disabledContentColor = KdsColors.actionDisabled
        )
    }
}

private fun getColorMain(color: KdsButtonColor): Color {
    return when (color) {
        KdsButtonColor.PRIMARY -> KdsColors.primaryMain
        KdsButtonColor.SECONDARY -> KdsColors.secondaryMain
        KdsButtonColor.SUCCESS -> KdsColors.successMain
        KdsButtonColor.WARNING -> KdsColors.warningMain
        KdsButtonColor.ERROR -> KdsColors.errorMain
    }
}

private fun getColorContrast(color: KdsButtonColor): Color {
    return when (color) {
        KdsButtonColor.PRIMARY -> KdsColors.primaryContrastText
        KdsButtonColor.SECONDARY -> KdsColors.secondaryContrastText
        KdsButtonColor.SUCCESS -> KdsColors.successContrastText
        KdsButtonColor.WARNING -> KdsColors.warningContrastText
        KdsButtonColor.ERROR -> KdsColors.errorContrastText
    }
}

private fun getButtonPadding(size: KdsButtonSize): PaddingValues {
    return when (size) {
        KdsButtonSize.SMALL -> PaddingValues(horizontal = 12.dp, vertical = 6.dp)
        KdsButtonSize.MEDIUM -> PaddingValues(horizontal = 16.dp, vertical = 10.dp)
        KdsButtonSize.LARGE -> PaddingValues(horizontal = 24.dp, vertical = 14.dp)
    }
}

private fun getButtonHeight(size: KdsButtonSize): Dp {
    return when (size) {
        KdsButtonSize.SMALL -> 36.dp
        KdsButtonSize.MEDIUM -> 44.dp
        KdsButtonSize.LARGE -> KdsDimensions.buttonMinHeight
    }
}

// =============================================================================
// PREVIEWS
// =============================================================================

@Preview(name = "Contained - Primary", showBackground = true)
@Composable
private fun PreviewContainedPrimary() {
    KdsTheme {
        KdsButton(
            text = "Pay Now",
            onClick = { },
            variant = KdsButtonVariant.CONTAINED,
            color = KdsButtonColor.PRIMARY
        )
    }
}

@Preview(name = "Contained - Success", showBackground = true)
@Composable
private fun PreviewContainedSuccess() {
    KdsTheme {
        KdsButton(
            text = "Confirm Payment",
            onClick = { },
            variant = KdsButtonVariant.CONTAINED,
            color = KdsButtonColor.SUCCESS
        )
    }
}

@Preview(name = "Outlined - Primary", showBackground = true)
@Composable
private fun PreviewOutlinedPrimary() {
    KdsTheme {
        KdsButton(
            text = "Cancel",
            onClick = { },
            variant = KdsButtonVariant.OUTLINED,
            color = KdsButtonColor.PRIMARY
        )
    }
}

@Preview(name = "Outlined - Error", showBackground = true)
@Composable
private fun PreviewOutlinedError() {
    KdsTheme {
        KdsButton(
            text = "Abandon Payment",
            onClick = { },
            variant = KdsButtonVariant.OUTLINED,
            color = KdsButtonColor.ERROR
        )
    }
}

@Preview(name = "Text Button", showBackground = true)
@Composable
private fun PreviewTextButton() {
    KdsTheme {
        KdsButton(
            text = "Learn More",
            onClick = { },
            variant = KdsButtonVariant.TEXT,
            fullWidth = false
        )
    }
}

@Preview(name = "Loading State", showBackground = true)
@Composable
private fun PreviewLoading() {
    KdsTheme {
        KdsButton(
            text = "Processing...",
            onClick = { },
            loading = true
        )
    }
}

@Preview(name = "Disabled State", showBackground = true)
@Composable
private fun PreviewDisabled() {
    KdsTheme {
        KdsButton(
            text = "Continue",
            onClick = { },
            enabled = false
        )
    }
}

@Preview(name = "Size Variants", showBackground = true)
@Composable
private fun PreviewSizes() {
    KdsTheme {
        androidx.compose.foundation.layout.Column {
            KdsButton(
                text = "Small",
                onClick = { },
                size = KdsButtonSize.SMALL,
                fullWidth = false
            )
            Spacer(Modifier.height(8.dp))
            KdsButton(
                text = "Medium",
                onClick = { },
                size = KdsButtonSize.MEDIUM,
                fullWidth = false
            )
            Spacer(Modifier.height(8.dp))
            KdsButton(
                text = "Large",
                onClick = { },
                size = KdsButtonSize.LARGE,
                fullWidth = false
            )
        }
    }
}

@Preview(name = "All Variants", showBackground = true)
@Composable
private fun PreviewAllVariants() {
    KdsTheme {
        androidx.compose.foundation.layout.Column(
            modifier = Modifier.fillMaxWidth()
        ) {
            KdsButton(
                text = "Contained",
                onClick = { },
                variant = KdsButtonVariant.CONTAINED
            )
            Spacer(Modifier.height(8.dp))
            KdsButton(
                text = "Outlined",
                onClick = { },
                variant = KdsButtonVariant.OUTLINED
            )
            Spacer(Modifier.height(8.dp))
            KdsButton(
                text = "Text",
                onClick = { },
                variant = KdsButtonVariant.TEXT
            )
        }
    }
}

File: ios/Sources/Components/KdsButton.swift
import SwiftUI

/**
 * Khipu Design System - Button Component
 *
 * A themed button component with variants (contained, outlined, text),
 * color schemes, sizes, and loading state support.
 *
 * Built with SwiftUI following Material Design principles.
 */

// MARK: - Button Variant

/// Visual style variant for KdsButton
public enum KdsButtonVariant {
    /// Filled button with solid background color
    case contained
    /// Button with border and transparent background
    case outlined
    /// Text-only button with no background or border
    case text
}

// MARK: - Button Color

/// Color scheme for KdsButton
public enum KdsButtonColor {
    case primary
    case secondary
    case success
    case warning
    case error
    case info
}

// MARK: - Button Size

/// Size variant for KdsButton
public enum KdsButtonSize {
    case small
    case medium
    case large

    var height: CGFloat {
        switch self {
        case .small: return 36
        case .medium: return 44
        case .large: return 56
        }
    }

    var horizontalPadding: CGFloat {
        switch self {
        case .small: return 12
        case .medium: return 16
        case .large: return 24
        }
    }

    var verticalPadding: CGFloat {
        switch self {
        case .small: return 6
        case .medium: return 10
        case .large: return 14
        }
    }

    var fontSize: CGFloat {
        switch self {
        case .small: return KdsTokens.Typography.fontSizeSm
        case .medium: return KdsTokens.Typography.fontSizeBase
        case .large: return KdsTokens.Typography.fontSizeLg
        }
    }
}

// MARK: - Button Component

/// Khipu Design System Button
///
/// A versatile button component with multiple variants, colors, and states.
///
/// **Example:**
/// ```swift
/// KdsButton("Pay Now") {
///     // Handle tap
/// }
/// .variant(.contained)
/// .color(.primary)
/// .loading(isProcessing)
///
/// KdsButton("Cancel", variant: .outlined, color: .error) {
///     // Handle tap
/// }
/// ```
public struct KdsButton: View {
    // MARK: - Properties

    private let title: String
    private let action: () -> Void
    private var variant: KdsButtonVariant
    private var color: KdsButtonColor
    private var size: KdsButtonSize
    private var isEnabled: Bool
    private var isLoading: Bool
    private var fullWidth: Bool
    private var leadingIcon: Image?
    private var trailingIcon: Image?

    // MARK: - Initializer

    /// Creates a button with the specified title and action
    ///
    /// - Parameters:
    ///   - title: The button label text
    ///   - variant: Visual style variant (default: .contained)
    ///   - color: Color scheme (default: .primary)
    ///   - size: Button size (default: .large)
    ///   - action: Callback when button is tapped
    public init(
        _ title: String,
        variant: KdsButtonVariant = .contained,
        color: KdsButtonColor = .primary,
        size: KdsButtonSize = .large,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.action = action
        self.variant = variant
        self.color = color
        self.size = size
        self.isEnabled = true
        self.isLoading = false
        self.fullWidth = false
        self.leadingIcon = nil
        self.trailingIcon = nil
    }

    // MARK: - Body

    public var body: some View {
        Button(action: {
            if !isLoading && isEnabled {
                action()
            }
        }) {
            HStack(spacing: 8) {
                if isLoading {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: contentColor))
                        .scaleEffect(0.8)
                } else {
                    if let icon = leadingIcon {
                        icon
                            .resizable()
                            .scaledToFit()
                            .frame(width: 20, height: 20)
                    }

                    Text(title)
                        .font(.system(size: size.fontSize, weight: .medium))
                        .lineLimit(1)

                    if let icon = trailingIcon {
                        icon
                            .resizable()
                            .scaledToFit()
                            .frame(width: 20, height: 20)
                    }
                }
            }
            .foregroundColor(contentColor)
            .frame(maxWidth: fullWidth ? .infinity : nil)
            .frame(height: size.height)
            .padding(.horizontal, size.horizontalPadding)
            .background(backgroundColor)
            .cornerRadius(KdsTokens.BorderRadius.button)
            .overlay(
                RoundedRectangle(cornerRadius: KdsTokens.BorderRadius.button)
                    .stroke(borderColor, lineWidth: borderWidth)
            )
        }
        .disabled(!isEnabled || isLoading)
        .opacity(isEnabled && !isLoading ? 1.0 : 0.6)
    }

    // MARK: - Color Helpers

    private var backgroundColor: Color {
        guard isEnabled || isLoading else {
            return variant == .contained ? KdsTokens.Colors.actionDisabledBackground : .clear
        }

        switch variant {
        case .contained:
            return mainColor
        case .outlined, .text:
            return .clear
        }
    }

    private var contentColor: Color {
        guard isEnabled || isLoading else {
            return KdsTokens.Colors.actionDisabled
        }

        switch variant {
        case .contained:
            return contrastColor
        case .outlined, .text:
            return mainColor
        }
    }

    private var borderColor: Color {
        guard isEnabled || isLoading else {
            return KdsTokens.Colors.actionDisabled
        }

        return variant == .outlined ? mainColor : .clear
    }

    private var borderWidth: CGFloat {
        return variant == .outlined ? 1 : 0
    }

    private var mainColor: Color {
        switch color {
        case .primary: return KdsTokens.Colors.primaryMain
        case .secondary: return KdsTokens.Colors.secondaryMain
        case .success: return KdsTokens.Colors.successMain
        case .warning: return KdsTokens.Colors.warningMain
        case .error: return KdsTokens.Colors.errorMain
        case .info: return KdsTokens.Colors.infoMain
        }
    }

    private var contrastColor: Color {
        switch color {
        case .primary: return KdsTokens.Colors.primaryContrastText
        case .secondary: return KdsTokens.Colors.secondaryContrastText
        case .success: return KdsTokens.Colors.successContrastText
        case .warning: return KdsTokens.Colors.warningContrastText
        case .error: return KdsTokens.Colors.errorContrastText
        case .info: return KdsTokens.Colors.infoContrastText
        }
    }
}

// MARK: - View Modifiers

public extension KdsButton {
    /// Sets the button variant
    func variant(_ variant: KdsButtonVariant) -> KdsButton {
        var button = self
        button.variant = variant
        return button
    }

    /// Sets the button color scheme
    func color(_ color: KdsButtonColor) -> KdsButton {
        var button = self
        button.color = color
        return button
    }

    /// Sets the button size
    func size(_ size: KdsButtonSize) -> KdsButton {
        var button = self
        button.size = size
        return button
    }

    /// Sets whether the button is enabled
    func enabled(_ isEnabled: Bool) -> KdsButton {
        var button = self
        button.isEnabled = isEnabled
        return button
    }

    /// Sets the loading state
    func loading(_ isLoading: Bool) -> KdsButton {
        var button = self
        button.isLoading = isLoading
        return button
    }

    /// Sets whether the button should fill the available width
    func fullWidth(_ fullWidth: Bool = true) -> KdsButton {
        var button = self
        button.fullWidth = fullWidth
        return button
    }

    /// Sets a leading icon
    func leadingIcon(_ icon: Image?) -> KdsButton {
        var button = self
        button.leadingIcon = icon
        return button
    }

    /// Sets a trailing icon
    func trailingIcon(_ icon: Image?) -> KdsButton {
        var button = self
        button.trailingIcon = icon
        return button
    }
}

// MARK: - Previews

#if DEBUG
struct KdsButton_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 16) {
            // Contained variants
            Group {
                Text("Contained Variants")
                    .font(.headline)

                KdsButton("Pay Now", variant: .contained, color: .primary) {}

                KdsButton("Confirm", variant: .contained, color: .success) {}

                KdsButton("Warning", variant: .contained, color: .warning) {}

                KdsButton("Delete", variant: .contained, color: .error) {}
            }

            Divider()

            // Outlined variants
            Group {
                Text("Outlined Variants")
                    .font(.headline)

                KdsButton("Cancel", variant: .outlined, color: .primary) {}

                KdsButton("Info", variant: .outlined, color: .info) {}
            }

            Divider()

            // Text variant
            Group {
                Text("Text Variant")
                    .font(.headline)

                KdsButton("Learn More", variant: .text, color: .primary) {}
                    .fullWidth(false)
            }

            Divider()

            // States
            Group {
                Text("States")
                    .font(.headline)

                KdsButton("Loading", variant: .contained, color: .primary) {}
                    .loading(true)

                KdsButton("Disabled", variant: .contained, color: .primary) {}
                    .enabled(false)
            }

            Divider()

            // Sizes
            Group {
                Text("Sizes")
                    .font(.headline)

                KdsButton("Small", variant: .contained, color: .primary, size: .small) {}
                    .fullWidth(false)

                KdsButton("Medium", variant: .contained, color: .primary, size: .medium) {}
                    .fullWidth(false)

                KdsButton("Large", variant: .contained, color: .primary, size: .large) {}
                    .fullWidth(false)
            }
        }
        .padding()
    }
}
#endif
